llvm/lib/CAS/ObjectStore.cpp

```diff
@@ -98,6 +99,48 @@ Expected<ObjectProxy> ObjectStore::createProxy(ArrayRef<ObjectRef> Refs,
   return getProxy(*Ref);
 }
 
+Expected<ObjectRef>
+ObjectStore::storeFromOpenFileImpl(sys::fs::file_t FD,
+                                   Optional<sys::fs::file_status> Status) {
+  // \p Copy the file into an immutable memory buffer and call \p store on that.
+  // Using \p mmap would be unsafe because there's a race window between when we
+  // get the digest hash for the \p mmap contents and when we store the data; if
+  // the file changes in-between we will create an invalid object.
+
+  // FIXME: For the on-disk CAS implementation use cloning to store it as a
+  // standalone file if the file-system supports it and the file is large.
+
+  constexpr size_t ChunkSize = 4 * 4096;
+  // FIXME: MSVC: SmallString<ChunkSize * 2>
+  SmallString<4 * 4096 * 2> Data;
+  if (Error E = sys::fs::readNativeFileToEOF(FD, Data, ChunkSize))
+    return std::move(E);
+  return store(std::nullopt, ArrayRef(Data.data(), Data.size()));
+}
+
+Error ObjectStore::validateTree(ObjectRef Root) {
+  SmallDenseSet<ObjectRef> ValidatedRefs;
+  SmallVector<ObjectRef, 16> RefsToValidate;
+  RefsToValidate.push_back(Root);
+
+  while (!RefsToValidate.empty()) {
+    ObjectRef Ref = RefsToValidate.pop_back_val();
+    auto [I, Inserted] = ValidatedRefs.insert(Ref);
+    if (!Inserted)
+      continue; // already validated.
+    if (Error E = validate(getID(Ref)))
+      return E;
+    Expected<ObjectHandle> Obj = load(Ref);
+    if (!Obj)
+      return Obj.takeError();
+    cantFail(forEachRef(*Obj, [&RefsToValidate](ObjectRef R) -> Error {
```

Comment:
This `cantFail` is suspicious: it's correct for our existing CAS implementations, but the documentation doesn't require that `forEachRef` will only fail if the callback fails.  We should either return the error here, or update the documentation on `forEachRef`.